File: cmd-tools/ts/fetch.ts
import https from "node:https"
import { performance } from "node:perf_hooks";

export async function fetchUrl(urls: string[]): Promise<void> {
    urls.forEach(url => {
        https.get(url, res => {
            let data = "";
            res.on("data", chunk => data += chunk);
            res.on("end", () => {
                console.log(data);
                console.log("Status: ", res.statusCode);
            });
        }).on("error", err => {
            console.error(`fetch: ${err}`);
            process.exit(1);
        })
    })
}

export async function fetchAll(urls: string[]) {
    const start = performance.now();

    const fetchUrl = (url: string): Promise<string> => {
        const start = performance.now();
        return new Promise((resolve) => {
            https.get(url, res => {
                let data = "";
                res.on("data", chunk => data += chunk);
                res.on("end", () => {
                    const secs = (performance.now() - start) / 1000;
                    resolve(`${secs.toFixed(2)}s ${data.length} ${url}`);
                });
            }).on("error", err => {
                resolve(`fetch: ${url}: ${err.message}`);
            })
        })
    }

    const results = urls.map(url => fetchUrl(url));
    for (const r of results) {
        console.log(await r);
    }
    const secs = (performance.now() - start) / 1000;
    console.log(`${secs.toFixed(2)}s elapsed`);
}

File: cmd-tools/python/fetch.py
import concurrent.futures
import sys
import time

import requests


def fetch(urls: list[str]):
    for url in urls:
        try:
            res = requests.get(url)
            print(res.text)
            print("Status: ", res.status)
        except requests.exceptions.HTTPError as e:
            print(f"fetch: HTTP error for {url}: {e}", file=sys.stderr)
            sys.exit(1)
        except requests.exceptions.RequestException as e:
            print(f"fetch: Error fetching {url}: {e}", file=sys.stderr)
            sys.exit(1)


def fetchall(urls: list[str]):

    def fetch_url(url: str):
        start = time.time()
        try:
            res = requests.get(url)
            nbytes = len(res.content)
            secs = time.time() - start
            return f"{secs:.2f}s {nbytes:7} {url}"
        except requests.exceptions.HTTPError as e:
            print(f"fetch: HTTP error for {url}: {e}", file=sys.stderr)
            sys.exit(1)
        except requests.exceptions.RequestException as e:
            print(f"fetch: Error fetching {url}: {e}", file=sys.stderr)
            sys.exit(1)

    start = time.time()
    with concurrent.futures.ThreadPoolExecutor() as executor:
        results = list(executor.map(fetch_url, urls))

    for r in results:
        print(r)

    print(f"{time.time() - start:.2f}s elapsed")
